logbook/oplog/log.go

```diff
@@ -92,6 +93,68 @@ type Logstore interface {
 	Descendants(ctx context.Context, l *Log) error
 }
 
+// SparseAncestorsAllDescendantsLogstore is a an extension interface to
+// Logstore with an optimized method for getting a log with sparse parents and
+// all descendants
+type SparseAncestorsAllDescendantsLogstore interface {
+	Logstore
+	// GetSparseAncestorsAllDescendants is a fast-path method for getting a
+	// log that includes sparse parents & complete descendants. "sparse parents"
+	// have the only children given in parent specified
+	// AllDescendants include
+	// the returned log will match the ID of the request, with parents
+	GetSparseAncestorsAllDescendants(ctx context.Context, id string) (*Log, error)
+}
+
+// GetWithSparseAncestorsAllDescendants is a fast-path method for getting a
+// log that includes sparse parents & complete descendants. "sparse parents"
+// means the only children given in parent
+// the returned log will match the ID of the request, with parents
+func GetWithSparseAncestorsAllDescendants(ctx context.Context, store Logstore, id string) (*Log, error) {
+	if sparseLS, ok := store.(SparseAncestorsAllDescendantsLogstore); ok {
+		return sparseLS.GetSparseAncestorsAllDescendants(ctx, id)
+	}
+
+	l, err := store.Get(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO (b5) - this is error prone
```

Comment:
Could you expand on this a bit? Do you mean its bad if this condition is true, and if so what might cause it? Or do you mean it's bad how this condition is being handled?

Reply:
I'll update that comment with this:

```
        // TODO (b5) - this is error prone. logs may exist but not be fetched. This
	// check is here now because not all implementations support the Descendants
	// method properly.
	// in the future remove this check once all implementations we know of have
	// a working `Descendants` implementation
```